runtime/lua/vim/shared.lua

```diff
@@ -396,6 +396,37 @@ function vim.tbl_deep_extend(behavior, ...)
   return tbl_extend(behavior, true, ...)
 end
 
+--- Traverses any tree-like object defined by `key`,
+--- calling the provided `visitor` on each descendant.
+---
+--- Example:
+---
+--- ```lua
+--- local comments = {}
+--- vim.traverse(
+---   language_tree,
+---   function(tree) return tree:children() end,
+---   function(tree)
+---     if tree:lang() == "comment" then
+---       table.insert(comments, tree)
+---     end
+---   end
+--- )
+--- ```
+---
+---@param tree any The object to traverse.
+---@param key string|fun(subtree: any): any[] Define the "children" to visit.
+---@param visitor fun(subtree: any): boolean|nil Function to call on each descendant.
+---       Return true to skip the rest of `subtree`.
```

Comment:
I think it would be helpful to return the `node` if `visitor` returns `true` (and else return `nil`). TypeScript does something like that [here](https://github.com/microsoft/TypeScript/blob/b3770e78527a379d2345412bbab5acc0bafe0090/src/compiler/core.ts#L34).

Reply:
Interesting! I thought about `vim.traverse` having a return value; what makes it hard is that you need to apply a reducer since multiple branches may return. Whereas when iterating over a list there is only one branch.

I could make a `tree_reduce` but I think I'd be getting ahead of myself.

I'll think about it some more!

Reply:
Hmmm that's true. There's also the hacky approach of having a special return value for indicating that the traversal must stop completely, like done [here](https://github.com/microsoft/TypeScript/blob/b3770e78527a379d2345412bbab5acc0bafe0090/src/compiler/utilitiesPublic.ts#L717). I don't like this too much though :/

Reply:
I think that makes sense for a "find" use-case rather than a "map"/"forEach" use-case; the set of things that gets iterated vs cancelled becomes unpredictable for the user even if they know (leaky abstraction) whether it's DFS, BFS, or iterated DFS

The intention of this PR was to serve the "forEach" use-case and internally we usually use recursion in that way (and so cancelling wants to be of children not of siblings). But I've seen at least one plugin on GitHub Search that used it in a "find"-y way, so who knows, maybe "find_in_tree" or something could be a future utility lol.

Reply:
Sounds reasonable :)